cmd/format-xl.go

```diff
@@ -443,6 +446,82 @@ func checkFormatXLValues(formats []*formatXLV3) error {
 	return nil
 }
 
+// Get Deployment ID for the XL sets from format.json.
+// This need not be in quorum. Even if one of the format.json
+// file has this value, we assume it is valid.
+// If more than one format.json's have different id, it is considered a corrupt
+// backend format.
+func formatXLGetDeploymentID(refFormat *formatXLV3, formats []*formatXLV3) (string, error) {
+	var deploymentID string
+	for _, format := range formats {
+		if format == nil || format.ID == "" {
+			continue
+		}
+		if reflect.DeepEqual(format.XL.Sets, refFormat.XL.Sets) {
+			// Found an ID in one of the format.json file
+			// Set deploymentID for the first time.
+			if deploymentID == "" {
+				deploymentID = format.ID
+			} else if deploymentID != format.ID {
+				// DeploymentID found earlier doesn't match with the
+				// current format.json's ID.
+				return "", errCorruptedFormat
+			}
+		}
+	}
+	return deploymentID, nil
+}
+
+// formatXLFixDeploymentID - Add deployment id if it is not present.
+func formatXLFixDeploymentID(ctx context.Context, storageDisks []StorageAPI, refFormat *formatXLV3) (err error) {
+	// Acquire lock on format.json
+	mutex := newNSLock(globalIsDistXL)
+	formatLock := mutex.NewNSLock(minioMetaBucket, formatConfigFile)
+	if err = formatLock.GetLock(globalHealingTimeout); err != nil {
+		return err
+	}
+	defer formatLock.Unlock()
+	// Attempt to load all `format.json` from all disks.
+	formats, sErrs := loadFormatXLAll(storageDisks)
+	for i, sErr := range sErrs {
+		if _, ok := formatCriticalErrors[sErr]; ok {
+			return fmt.Errorf("Disk %s: %s", globalEndpoints[i], sErr)
+		}
+	}
+	refFormat.ID, err = formatXLGetDeploymentID(refFormat, formats)
+	if err != nil {
+		return err
+	}
+
+	// If ID is generated for the first time,
+	// We set the ID in all the formats and update.
+	if refFormat.ID == "" {
+		refFormat.ID = mustGetUUID()
+		for _, format := range formats {
+			if format != nil {
+				format.ID = refFormat.ID
+			}
+		}
+		// Deployment ID needs to be set on all the disks.
+		// Save `format.json` across all disks.
+		if err = saveFormatXLAll(ctx, storageDisks, formats); err != nil {
+			return err
+		}
+	}
+
+	// Update if deploymentID is empty on a local disk
+	for index, format := range formats {
+		if format != nil && format.ID == "" && globalEndpoints[index].IsLocal {
+			format.ID = refFormat.ID
+			if err = saveFormatXL(storageDisks[index], format); err != nil {
```

Comment:
Better to use logger package here to log (to know the line number where it failed)